src/Microsoft.Identity.Web/CookiePolicyOptionsExtensions.cs

```diff
@@ -9,175 +9,175 @@
+		/// <param name="options"><see cref="CookiePolicyOptions"/>to update.</param>
+		/// <param name="disallowsSameSiteNone">If you don't want to use the default user agent list implementation,
+		/// the method sent in this parameter will be run against the user agent and if returned true, SameSite value will be set to Unspecified.
+		/// The default user agent list used can be found at: https://devblogs.microsoft.com/aspnet/upcoming-samesite-cookie-changes-in-asp-net-and-asp-net-core/. </param>
+		/// <returns><see cref="CookiePolicyOptions"/> to chain.</returns>
+		public static CookiePolicyOptions HandleSameSiteCookieCompatibility(this CookiePolicyOptions options, Func<string, bool> disallowsSameSiteNone)
+		{
+			if (options == null)
+			{
+				throw new ArgumentNullException(nameof(options));
+			}
+
+			options.MinimumSameSitePolicy = SameSiteMode.Unspecified;
+			options.OnAppendCookie = cookieContext =>
+				CheckSameSite(cookieContext.Context, cookieContext.CookieOptions, disallowsSameSiteNone);
+			options.OnDeleteCookie = cookieContext =>
+				CheckSameSite(cookieContext.Context, cookieContext.CookieOptions, disallowsSameSiteNone);
+
+			return options;
+		}
+
+		private static void CheckSameSite(HttpContext httpContext, CookieOptions options, Func<string, bool> disallowsSameSiteNone)
+		{
+			if (options.SameSite == SameSiteMode.None)
+			{
+				var userAgent = httpContext.Request.Headers[Constants.UserAgent].ToString();
+				if (disallowsSameSiteNone(userAgent))
+				{
+					options.SameSite = SameSiteMode.Unspecified;
+				}
+			}
+		}
+
+		/// <summary>
+		/// Checks if the specified user agent supports "SameSite=None" cookies.
+		/// </summary>
+		/// <param name="userAgent">Browser user agent.</param>
+		/// <remarks>
+		/// Incompatible user agents include:
+		/// <list type="bullet">
+		/// <item>Versions of Chrome from Chrome 51 to Chrome 66 (inclusive on both ends).</item>
+		/// <item>Versions of UC Browser on Android prior to version 12.13.2.</item>
+		/// <item>Versions of Safari and embedded browsers on MacOS 10.14 and all browsers on iOS 12.</item>
+		/// </list>
+		/// Reference: https://www.chromium.org/updates/same-site/incompatible-clients.
+		/// </remarks>
+		/// <returns>True, if the user agent does not allow "SameSite=None" cookie; otherwise, false.</returns>
+		public static bool DisallowsSameSiteNone(string userAgent)
+		{
+			return HasWebKitSameSiteBug() ||
+				DropsUnrecognizedSameSiteCookies();
+
+			bool HasWebKitSameSiteBug() =>
+				IsIosVersion(12) ||
+				(IsMacosxVersion(10, 14) &&
+				(IsSafari() || IsMacEmbeddedBrowser()));
+
+			bool DropsUnrecognizedSameSiteCookies()
+			{
+				if (IsUcBrowser())
+				{
+					return !IsUcBrowserVersionAtLeast(12, 13, 2);
+				}
+
+				return IsChromiumBased() &&
+					IsChromiumVersionAtLeast(51) &&
+					!IsChromiumVersionAtLeast(67);
+			}
+
+			bool IsIosVersion(int major)
+			{
+				string regex = @"\(iP.+; CPU .*OS (\d+)[_\d]*.*\) AppleWebKit\/";
```

Comment:
This code is in the hotpath of requests. As such, it would likely speed up by 5-7x if regular expressions were avoided where possible. We did a similar change in a planet scale service and sped the code up by 7x.